### DISCUSSION TITLE
Where help is needed most urgently?

### DISCUSSION THREAD

**Discussion by a1exsh** [ihhub/fheroes2#4071] [General]
Hi!

I'm new to the project, but I'm a long-standing fan of Heroes II.  So first of all — thanks a ton for making the dream come true! :)

I'm a professional software engineer with 15+ years of experience, though, to be honest, I haven't touched C++ for more than 10 years now. ;)

I'd really like to help with the development, but I need some guidance: mainly, where help is needed the most?

I understand that fixing bugs should have higher priority than adding features.  Should I also try to take some of the issues marked with `help wanted`?

Should I first try to help with review of open PRs rather than opening more of them? :)

One alarming sign for me is complete(?) lack of automated tests.  Is this problem generally recognized by the project?  I think it makes sense to invest into tests so that we don't fix issues twice.  In my experience introducing tests also leads to rewriting code in a cleaner way as a side-effect. ;)

Cheers!

Alex

**Comment by ihhub** [comment]
Welcome, Alex!

Thanks again for your contribution to the project!

We have multiple areas to dig in and I'll try to explain issues we're facing:
- Translation to other languages. Perhaps a straightforward task with almost zero development knowledge required. The initial aim to add support for official languages such as English (it's done by default), German, Polish and Russian. Only these 4 languages we're aware of being used in official HoMM2 distributions. We have couple of issues in this area but not much. In this future we have to find a way to generate other language alphabets from English sprites (images). We can't distribute any original game resources as it's against licensing.
- AI. It is done by around 10%. Some things are in real mess due to leftovers from previous developers but overall we have to put a lot of code logic to make AI really smart. Talking about machine learning and other interesting things are too early as the basic logic is not there. You can find many issues related to AI and how to improve it. Some ideas are posted as well. This is a bit shady area as sometimes nobody can give you the concrete answer how AI should behave here and there but common logic must exist. I highly recommend to start from small issues if you want to explore such area of interest.
- Game logic. The most important and at the same time very hard to fix. It could be divided 3 main subgroups: battle logic, adventure map logic and main game logic. The first one requires polishing as something is not implemented yet, like for example straight attacks of wide creatures. Adventure map logic is usually related to passability of objects. This is a head cracking area which we're still figuring out how it works in the original game. I'd say this is the most difficult but the most critical at the moment. Game logic is logic related to artifacts, monsters, diplomacy, special campaign logic and etc. Also this is related to experimental options which were added by the previous team.
- Sound and music. This area is almost stable. We have few OS related things but overall look fine.
- Video. Just a couple of issues related to it which is low priority for now.
- Graphics or UI. Most of open issues related to it. One the biggest problem is proper rendering of objects of adventure map. You can find few pull request trying to fix problems related to it and one WIP pull request contains totally new logic for rendering. We have a dozen of issues related to it. Somehow it might be related to game logic: the order of drawing. Besides this we have multiple UI issues on position of elements in windows, the order of drawing during battle and some glitches from the original game and from our new rendered stuff.

I hope now you have a broader picture what's going on and where the help is required. On the topic of pull request you're encouraged to put comments if you have any valid and useful. Everyone is allowed to do such. Every pull request is reviewed at least by one person (by me). So if pull requests are not reviewed over time just poke me again or any other collaborator to review it as people here might work on multiple things and just forget to do it.

We had conversation about unit tests in the past. The biggest problem with them is maintenance and also UI verification. If you're willing to take this step you can take a look at [this code](`https://github.com/ihhub/penguinV/blob/master/test/unit_tests/unit_test_framework.cpp`) as a basis. Ideally, we shouldn't use third party integration inside the source code to keep as less dependencies as possible. I think this is a huge topic to cover so to be short: yes, we're aware of unit tests absence :) Please keep in mind that a big chunk of code is still from the previous team and we're trying to deal with it.

Furthermore you can join our Discord channel to discuss any other development topics.

**Reply by pflugs30** [reply]
@a1exsh I wish I would have seen this discussion when I first joined the team!  My introduction is almost exactly like yours, down to the years of experience.

I know you've done a lot of testing based on your comments.  What areas have you pushed into for coding?  I've tried my hand at a few simple UI and events loop fixes.

I've also noticed a lot of issues popping up regarding passability.  I personally feel that those are some of the highest priority bugs to fix as they prevent the user from doing what they want (or reproducing behavior in the OG).  AI decision making would also be important with a priority on OG reproduction.

If there's an area of code you'd like to tag-team, I'm all for working together on some stuff, too (i.e., pair programming).

**Reply by a1exsh** [reply]
@pflugs30 
> I know you've done a lot of testing based on your comments. What areas have you pushed into for coding? I've tried my hand at a few simple UI and events loop fixes.

This has been my focus recently as well: making buttons and shows pixel-perfect, and adding all the missing hot-keys ;)  However I've touched at least one place related to game logic, that is capture of artifacts.  From that experience I'm even more convinced that automated testing is a must — there are just too many ways it can break. :sweat_smile: 

> I've also noticed a lot of issues popping up regarding passability. I personally feel that those are some of the highest priority bugs to fix as they prevent the user from doing what they want (or reproducing behavior in the OG). AI decision making would also be important with a priority on OG reproduction.

In the past weeks I've reviewed all the issues marked as "bug" to get an overall impression and also picked up and fixed some of the low-hanging fruits (still keeping some for newcomers after me).  I didn't touch any adventure map rendering, as I understand fixing them one by one is pointless, given the new rendering code is already being worked on.  I also don't dare touching passability, but do I think we need to codify existing behavior in a regression test suite, so that fixing one passability issue doesn't inadvertently introduces two more.

> If there's an area of code you'd like to tag-team, I'm all for working together on some stuff, too (i.e., pair programming).

I do find pair programming invaluable, though not sure we can really make it work in a fully remote open source setup.  I'm in Central Europe, when it comes to Time Zones — and you? :)

**Reply by ihhub** [reply]
Hi @a1exsh and @pflugs30 , nobody expects from you to pick the most urgent tasks, We have a plenty of them so focus on those which you like to fix :)

If you have any questions and really want to understand how some parts of code work we might have a video session some time in the future.

**Reply by oleg-derevenetz** [reply]
@a1exsh @pflugs30 Passability issues are very complex because rules are unclear (there is no complete description anywhere) and also tile passability depends not only on the tile contents itself, but on the nature of the surrounding tiles.

**Reply by pflugs30** [reply]
@a1exsh 
> I do find pair programming invaluable, though not sure we can really make it work in a fully remote open source setup. I'm in Central Europe, when it comes to Time Zones — and you? :)

Central... US.  :-)

> If you have any questions and really want to understand how some parts of code work we might have a video session some time in the future.

Now THIS gets me excited!  I think this would be tremendously helpful and would be a great way to get some of us newcomers more familiar and comfortable with the code.  I'm imagining a webinar style-session where folks like @ihhub, @oleg-derevenetz , and @Branikolog (and others whom I've missed) talk through things like:

1. Code locations (i.e. where to find things like passability, AI for battle & movement, graphics).
2. How to work with existing resources like sprites
3. Coding style
4. Dev environment setup
5. Where to find resources in the Github project or Discord, etc.
6. Testing best practices
7. Other ideas submitted to a discussion thread prior to the webinar

The session could be recorded so that others in the future or opposite time zones could still watch and benefit.  We wouldn't solve any specific issues in the session, but we could cover a lot of ground in a short time.  I'd volunteer to coordinate and moderate something, though I'd also gladly give it up to a more senior member, too.

Tagging a few others I've spoken to recently to notify them: @zenseii @LeHerosInconnu

**Reply by a1exsh** [reply]
@oleg-derevenetz 
> Passability issues are very complex because rules are unclear (there is no complete description anywhere) and also tile passability depends not only on the tile contents itself, but on the nature of the surrounding tiles.

That's to be expected :)  But the point I'm trying to make is that currently all the knowledge of these details and quirks are mostly in the head's of people who worked on these issues so far.  A sustainable approach would be to write these details down, preferrably in form of automated tests (but documentation could also be a good start), so that anyone can quickly check if their idea of fixing an issue doesn't break other corner cases.

**Reply by pflugs30** [reply]
@ihhub Any thoughts on the video session intro for new developers and contributors?  I think that could be extremely helpful in getting new folks up to speed more quickly.  If we do it well, then we can make the content available for future contributors, too.  If nothing else, it would be great to collaborate.

Again, I'm happy to help facilitate.

**Reply by ihhub** [reply]
Hi @pflugs30 ,

I think it'd be great to come up with the list of topics and questions which you want to clarify to make the session more efficient. From my side I'll try to prepare some extra information. Also we have to keep in mind that we have big time difference as all of us are spread across the world :)
